### ISSUE TITLE
Type statsd.timed

### ISSUE THREAD

**Issue by Meemaw** [DataDog/datadogpy#797]
**Note:** 
If you have a feature request, you should [contact support](https://docs.datadoghq.com/help/) so the request can be properly tracked.

**Is your feature request related to a problem? Please describe.**
`@statsd.timed` decorator is missing type hints, which results in a decorated function losing all of its type information.

https://github.com/DataDog/datadogpy/blob/master/datadog/dogstatsd/base.py#L801

**Describe the solution you'd like**
`@statsd.timed` decorator provides proper type hints, which correctly preserver the decorated function argument types.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

**Comment by drichardson** [comment]
Yes please. Whatnot is running into this now as well and are resorting to silly things like Sub-classing DogStatsd to workaround it. I was about to open up an issue with typeshed to add types (https://github.com/python/typeshed/pull/10843) but it got closed because supposedly datadogpy already supports types. Sure enough, I saw that some very basic typing is supported which is nice, but the problem with statsd.timed is really bad because it erases types we have on our own functions. That is, it does not only lack type hints itself, but it essentially removes them from our own code as well.

**Comment by drichardson** [comment]
I've submitted a support ticket about this issue on behalf of Whatnot here https://help.datadoghq.com/hc/en-us/requests/1381023

**Comment by drichardson** [comment]
Here is our current workaround.

```
from collections.abc import Callable
from typing import ParamSpec
from typing import TypeVar

from datadog import DogStatsd as _DogStatsd

P = ParamSpec("P")
T = TypeVar("T")


# This is a subclass of DogStatsd to make sure that all related decorators
# preserve the function type. The built in decorators will remove the function type
# and any type error will not be caught by mypy.
class DogStatsd(_DogStatsd):
    def timed(
        self,
        metric: str | None = None,
        tags: list[str] | None = None,
        sample_rate: float | None = None,
        use_ms: bool | None = None,
    ) -> Callable[[Callable[P, T]], Callable[P, T]]:
        return super().timed(
            metric=metric, tags=tags, sample_rate=sample_rate, use_ms=use_ms
        )
```
